### ISSUE TITLE
Sandstorm Support

### ISSUE THREAD

**Issue by timshannon** [keeweb/keeweb#390]
No issue, just a small request for adding support for running on Sandstorm.io.  It's fairly easy to do, and you shouldn't actually need to change any code to support it.

**Comment by kasoban** [comment]
@JamborJan was working on that at some point, see here for example:  #172  /  [JamborJan/keeweb-sandstorm](https://github.com/JamborJan/keeweb-sandstorm)

No idea about the current state and/or plans for that project though...

**Comment by antelle** [comment]
> you shouldn't actually need to change any code to support it

So, what should be done? As far as I can understand, there's some code to write, e.g. working with files, etc...

**Comment by timshannon** [comment]
@antelle as I understand it, sandstorm is just a platform where your apps are containers running linux instances.  Since keeweb runs basically entirely client side, you just need to follow the instructions here: https://docs.sandstorm.io/en/latest/developing/raw-pure-client-apps/

However it looks like here https://github.com/JamborJan/keeweb-sandstorm/issues/2 @jamborJan has taken it one step further and made the storage persist on the linux container as opposed to (correct me if I'm wrong), storing it purely on the client side in browser.

**Comment by JamborJan** [comment]
When you want a client only app: that's super easy, it's already working. But I'm following a use case where a team is sharing a KeeWeb with passwords and local storage makes no sense here. I would like to share access with others on my sandstorm server. That's the story behind it.

I know that KeeWeb has improved a lot and I was simply not able to take a closer look during the last time. I will catch up again in the next weeks and see what I can do.

**Comment by antelle** [comment]
@timshannon yes, that's what I mean, I think, apps in sandstorm must follow platform guides and save files to the same storage, as other apps.
@JamborJan yes, config has been improved, you can find a reference here: https://github.com/keeweb/keeweb/wiki/Configuration#json-app-config
According to Sandstorm guidelines, the app must represent a file, without open/load functionality. So, the config should be like this:
```json
{
    "canOpen": false,
    "canOpenDemo": false,
    "canOpenSettings": false,
    "canCreate": false,
    "canImportXml": false,
    "dropbox": false,
    "webdav": false,
    "gdrive": false,
    "onedrive": false,
    "files": [{
        "storage": "webdav",
        "name": "My file",
        "path": "webdav-url",
        "options": { "user": "", "password": "" }
    }]
}
```
This config should be generated by server, and the app must have meta-tag `kw-config` set to config.json. Server must also expose files as WebDAV (not sure, if it's already enabled in sandstorm).

**Comment by JamborJan** [comment]
Hey @antelle,

when I want to use my config file I have a problem with the meta tag. I changed ```index.html``` to have this line:

```
<meta name="kw-config" content="config.json">
```

But I get an error message:

```
Error loading app
There was an error loading app settings. Please double check app url or contact your administrator.
```

This one is different to the error messages I get when I'm typing a completely wrong location, then I see the same error message but get a log which is like this:

```
2016/10/21 14:03:26 [error] 12#0: *3 open() "/opt/app/keeweb/config.jsonjht" failed (2: No such file or directory), client: 127.0.0.1, server: localhost, request: "GET /config.jsonjht HTTP/1.1", host: "888e57893601.local.sandstorm.io:6080"
```

Is there a bug or am I doing something wrong?

Second question: is there no way that I can write the file directly to the file system? Sandstorm has a concept where I can write anything to /var/ and it would be super easy to put the kdbx file there instead of using webdav or anything else.

Thanks!

**Comment by JamborJan** [comment]
Hi @antelle 

Thanks for the hints. The first issue was that I didn't setup the ```config.json``` file, the error message in the browsers console was:

```
[Error] 2016-10-27T05:18:07.717Z [app] Invalid app config, no settings section
Object
```

I was lazy here and only copy [the example](https://github.com/keeweb/keeweb/issues/390#issuecomment-250072078) above which is missing the settings section and did not double check with the manuals you have send.

After I have fixed that and now this part is working. 

How can we deal with the thing about writing to the file system?

**Comment by antelle** [comment]
Oh, yes, sorry, I forgot about the comment, I should've mentioned that it got outdated.
I'm not sure about file system, at the moment it would be tough, to replace launcher. What other functions do we need to override? Maybe clipboard management, etc...

**Comment by JamborJan** [comment]
Right now I think there are these issues left:

## storage
- some kind of storage inside the grain (tiny sandbox Sandstorm uses)
- can we do something with localstorage 

## save button 
- save button in settings doesn't work right now with this error
- might be related to Sandstorm specific stuff
- might be related to the storage topic 

```
[Error] 2016-10-27T05:41:06.378Z [storage-cache] Error opening indexed db
SecurityError (DOM Exception 18): IDBFactory.open() called in an invalid security context
open
initDb — 30831eb58dc357f6a8e54f347cdbfb38.local.sandstorm.io:18:20254
save — 30831eb58dc357f6a8e54f347cdbfb38.local.sandstorm.io:18:20637
(anonyme Funktion) — 30831eb58dc357f6a8e54f347cdbfb38.local.sandstorm.io:18:12272
(anonyme Funktion) — 30831eb58dc357f6a8e54f347cdbfb38.local.sandstorm.io:15:25059
(anonyme Funktion) — 30831eb58dc357f6a8e54f347cdbfb38.local.sandstorm.io:7:31221
(anonyme Funktion) — 30831eb58dc357f6a8e54f347cdbfb38.local.sandstorm.io:8:3562
[native code]
	error (30831eb58dc357f6a8e54f347cdbfb38.local.sandstorm.io:13:26741)
	initDb (30831eb58dc357f6a8e54f347cdbfb38.local.sandstorm.io:18:20523)
	save (30831eb58dc357f6a8e54f347cdbfb38.local.sandstorm.io:18:20637)
	(anonyme Funktion) (30831eb58dc357f6a8e54f347cdbfb38.local.sandstorm.io:18:12272)
	(anonyme Funktion) (30831eb58dc357f6a8e54f347cdbfb38.local.sandstorm.io:15:25059)
	(anonyme Funktion) (30831eb58dc357f6a8e54f347cdbfb38.local.sandstorm.io:7:31221)
	(anonyme Funktion) (30831eb58dc357f6a8e54f347cdbfb38.local.sandstorm.io:8:3562)
	(anonyme Funktion)
```

## save as button
- doesn't work at the moment
- this is totally a Sandstorm related issue

```
[Error] Refused to load data:attachment/file;base64,A9mimmf7S7UBAAMAAhAAMcHy5r9xQ1C+WAUhavxa/wMEAAEAAAAEIADNHEY9Uk9BeUdnBtGOz0d2AxPv399Mu21WDIrn3VRYVAUgAEmimYxNV0RKH7gQ5bcx9ozJPearD7bPrWg5INs3nZuuBggAcBcAAAAAAAAHEAD3Khebci+9UVLc3WKxd6eSCCAAGDzjp3MLpYQ4Q7BrV/MKW0VPk 

...

21nDPRehwim/LyC8Xv4FirRqLertt9pATm2aY9WUgPWCETf6q4f8v288Rp0j1xw77CufjI+65qqlahAOyJ1sMwQPomOD99TkjsPwxaoYNt369bv9zwk86EWWMKaULF66RHSvzWHN/7iVYAi1+QJKmFTTcRcK0QF8nWfRkczpv9hYb9rw9JOkh+QK1yyn1I+QR45M+GLziRqbe89U95Ay0434DzUnNA== because it does not appear in the frame-src directive of the Content Security Policy.
```

**Comment by antelle** [comment]
IndexedDb issue looks weird. Can IndexedDb be used in Sandstorm?

**Comment by JamborJan** [comment]
Here is an issue about that in Sandstorm: https://github.com/sandstorm-io/sandstorm/issues/91#issuecomment-62223883

Any comment or idea from your side according to the essence of the threat:

> just freezing the object and forcing apps to use getItem/setItem sounds like the right approach.

**Comment by ricricucit** [comment]
@JamborJan is it actually possible to use keeweb for a team? (I'm specifically looking at keeweb + nextcloud integration)

**Comment by JamborJan** [comment]
Hey @ricricucit , sorry for the late reply. I'm not 100% sure but I guess the app is not made for concurrent edits when I get this issue right: https://github.com/keeweb/keeweb/issues/367

As I currently have no Nextcloud instance ready for testing I cannot check it myself. 

Furthermore I'll have to say that I cannot provide the Sandstorm app for now. I'm not able to dedicate the required time to this. Sorry. Everything I've done so far is available here: https://github.com/JamborJan/keeweb-sandstorm. Anybody who is able to help is welcome.
